pdns/lua-record.cc

```diff
@@ -60,30 +60,133 @@ class IsUpOracle
+  {
+    try {
+      string useragent = productName();
+      if (cd.opts.count("useragent")) {
+        useragent = cd.opts.at("useragent");
+      }
+      MiniCurl mc(useragent);
+
+      string content;
+      if(cd.opts.count("source")) {
+        ComboAddress src(cd.opts.at("source"));
+        content=mc.getURL(cd.url, &cd.rem, &src);
+      }
+      else {
+        content=mc.getURL(cd.url, &cd.rem);
+      }
+      if(cd.opts.count("stringmatch") && content.find(cd.opts.at("stringmatch")) == string::npos) {
+        throw std::runtime_error(boost::str(boost::format("unable to match content with `%s`") % cd.opts.at("stringmatch")));
+      }
+      if(!status) {
+        g_log<<Logger::Warning<<"LUA record monitoring declaring "<<cd.rem.toString()<<" UP for URL "<<cd.url<<"!"<<endl;
+      }
+      setUp(cd);
+    }
+    catch(std::exception& ne) {
+      if(status || first)
+        g_log<<Logger::Warning<<"LUA record monitoring declaring "<<cd.rem.toString()<<" DOWN for URL "<<cd.url<<", error: "<<ne.what()<<endl;
+      setDown(cd);
+    }
+  }
+  void checkTCP(const CheckDesc& cd, const bool status, const bool first = false) {
+    try {
+      Socket s(cd.rem.sin4.sin_family, SOCK_STREAM);
+      ComboAddress src;
+      s.setNonBlocking();
+      if (cd.opts.count("source")) {
+        src = ComboAddress(cd.opts.at("source"));
+        s.bind(src);
+      }
+      s.connect(cd.rem, 1);
+      if (!status) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" ";
+        if(cd.opts.count("source"))
+          g_log<<"(source "<<src.toString()<<") ";
+        g_log<<"UP!"<<endl;
+      }
+      setUp(cd);
+    }
+    catch (const NetworkError& ne) {
+      if(status || first) {
+        g_log<<Logger::Warning<<"Lua record monitoring declaring TCP/IP "<<cd.rem.toStringWithPort()<<" DOWN: "<<ne.what()<<endl;
+      }
+      setDown(cd);
+    }
+  }
 
-  struct Checker
+  void checkThread()
   {
-    std::thread thr;
-    bool status;
+    while (true)
+    {
+      std::chrono::system_clock::time_point checkStart = std::chrono::system_clock::now();
+      std::vector<statuses_t::iterator> toDelete;
+      {
+        ReadLock lock{&d_lock};
+        for (auto it = d_statuses.begin(); it != d_statuses.end(); ++it) {
+          auto& desc = it->first;
+          auto& state = it->second;
+
+          if (desc.url.empty()) { // TCP
+            checkTCP(desc, state->status, state->first);
+          } else { // URL
+            checkURL(desc, state->status, state->first);
+          }
+          if (state->first) {
+            state->first = false;
+          }
+          if (std::chrono::system_clock::from_time_t(state->lastAccess) < (checkStart - std::chrono::hours(1))) {
+            toDelete.push_back(it);
+          }
+        }
+      }
+      if (!toDelete.empty()) {
+        WriteLock lock{&d_lock};
```

Comment:
We release the lock after scanning the map and the actual removals, so I believe another thread might be able to insert an entry during that time. Since `d_statuses` is based on a `std::map` the stored iterators should still be valid, but that might at the very least warrant a comment.